src/main.rs

```diff
@@ -34,321 +33,323 @@ extern crate xi_core_lib;
-        let mut state = self.state.borrow_mut();
         match method {
-            "update" => state.edit_view.apply_update(&params["update"]),
-            "scroll_to" => state.edit_view.scroll_to(params["line"].as_u64().unwrap() as usize),
+            "update" => self.send_view_cmd(EditViewCommands::ApplyUpdate(params["update"].clone())),
+            "scroll_to" => self.send_view_cmd(EditViewCommands::ScrollTo(params["line"].as_u64().unwrap() as usize)),
             "available_themes" => (), // TODO
             "available_plugins" => (), // TODO
+            "available_languages" => (), // TODO
             "config_changed" => (), // TODO
+            "language_changed" => (), // TODO
             _ => println!("unhandled core->fe method {}", method),
         }
-        // TODO: edit view should probably handle this logic
-        self.invalidate();
-    }
-
-    pub fn invalidate(&self) {
-        self.handle.borrow().invalidate();
     }
 }
 
-fn create_main(core: Core) -> Result<WindowHandle, Error> {
-    let main_state = MainWinState::new();
-    let main_win = Rc::new(MainWin::new(core, main_state));
-    let main_win_handler = MainWinHandler {
-        win: main_win,
-    };
-
-    let menubar = menus::create_menus();
-
-    let mut builder = WindowBuilder::new();
-    builder.set_handler(Box::new(main_win_handler));
-    builder.set_title("xi-editor");
-    builder.set_cursor(Cursor::IBeam);
-    builder.set_menu(menubar);
-    let window = builder.build().unwrap();
-    Ok(window)
-}
-
 #[derive(Clone)]
-struct MyHandler {
-    runloop: RunLoopHandle,
-    win_handle: Arc<Mutex<Option<IdleHandle>>>,
+struct AppDispatcher {
+    app: Arc<Mutex<Option<App>>>,
 }
 
-impl MyHandler {
-    fn new(runloop: RunLoopHandle) -> MyHandler {
-        MyHandler {
-            runloop,
-            win_handle: Default::default(),
+impl AppDispatcher {
+    fn new() -> AppDispatcher {
+        AppDispatcher {
+            app: Default::default(),
         }
     }
+
+    fn set_app(&mut self, app: &App) {
+        *self.app.lock().unwrap() = Some(app.clone());
+    }
+
+    fn set_menu_listeners(&self, state: &mut UiState) {
+        let app = self.app.clone();
+        state.set_command_listener(move |cmd, mut ctx| {            
+            match cmd {
+                cmd if cmd == MenuEntries::Exit as u32 => {
+                    ctx.close();
+                }
+                cmd if cmd == MenuEntries::Open as u32 => {
+                    if let Some(app) = app.lock().unwrap().as_ref() {
+                        let filename = ctx.file_dialog(FileDialogType::Open, FileDialogOptions::default());
+                        if filename.is_err() {
+                            return;
+                        }
+                        let filename = filename.unwrap().into_string();
+                        if filename.is_err() { // invalid unicode data
+                            return;
```

Comment:
Printing something would be nice as a stopgap, and of course we need a real status mechanism.